Update slugify method for Azure DevOps TOC generation to be more readable

PHLUNK opened the pull request:
Hi yzhang,

I have updated the slugify method for Azure DevOps used when generating/updating a TOC in a markdown file. For most headers, it will keep working, but simultaneously non-UTF8-encoded readable slugs. I have read https://github.com/yzhang-gh/vscode-markdown/issues/802, and unfortunately, TOC entries starting with numbers remain problematic and are generally rendered as references to work items in Azure DevOps markdown render. Therefore, the original slugify method is only used to fully encode these using the original slugify method, which produces working references, albeit far-less readable for us humans.

* Perform partial URI encoding for non-number indexed TOC items.
* Perform original method if TOC entry starts with number.

yzhang-gh commented:
Hi, Thanks for your contributions.

I don't use Azure DevOps myself so I don't know what "slugify results" would work for most of the Azure DevOps users. But it will be better if you can add some test cases specifically for this PR, just like

https://github.com/yzhang-gh/vscode-markdown/blob/79fec4f8a1aacd33cc2126ff28a5aecff1276d61/src/test/suite/unit/slugify.test.ts#L10-L15

I would be happy to merge this in as long as its behavior is consistent.

PHLUNK commented:
Hi yzhang,

Thanks for getting back to me and your pointers for adding some tests to verify the adapted slugify method for Azure DevOps. I have added a couple to test the encoding of symbols under different scenarios of usage. I have also updated the expected test results of the original test instances for this method as not every string is now fully encoded.

All tests are passed successfully here and I have verified that the test results of this new function do result in working slugs in Azure DevOps.

Hope this helps and thanks again for having a look at my adaptations.

yzhang-gh commented:
Nice work. Thanks a lot 🍻.
